File: src/main/weather.ts
/**
 * Weather tool backed by free APIs (no API key required):
 *
 * - Zippopotam.us for US ZIP → place + lat/lon geocoding.
 * - Open-Meteo for current conditions + daily forecast.
 *
 * Kept deliberately small: input is a US ZIP (or the user's Config default),
 * output is a single JSON blob with current + next few days so the model can
 * phrase whatever the user asked for ("is it raining", "what's the high",
 * "do I need a jacket", etc.) without needing a second call.
 */

const ZIPPOPOTAM_URL = "https://api.zippopotam.us/us";
const OPEN_METEO_FORECAST_URL = "https://api.open-meteo.com/v1/forecast";
const FETCH_TIMEOUT_MS = 15_000;

export interface WeatherLocation {
  zip: string;
  place: string;
  state: string;
  lat: number;
  lon: number;
}

export interface WeatherCurrent {
  time: string;
  temp_f: number;
  apparent_f: number;
  humidity_pct: number;
  wind_mph: number;
  wind_gusts_mph: number;
  precipitation_in: number;
  weather: string;
  is_day: boolean;
}

export interface WeatherDaily {
  date: string;
  weather: string;
  high_f: number;
  low_f: number;
  precip_chance_pct: number;
  precip_sum_in: number;
  sunrise: string;
  sunset: string;
}

export interface WeatherResultPayload {
  units: "imperial";
  location: WeatherLocation;
  current: WeatherCurrent;
  daily: WeatherDaily[];
  error?: string;
}

export interface WeatherErrorPayload {
  error: string;
  zip?: string;
}

function normalizeZip(raw: string): string | null {
  const digits = raw.replace(/\D+/g, "");
  if (digits.length < 5) return null;
  return digits.slice(0, 5);
}

async function fetchJson(url: string): Promise<unknown> {
  const res = await fetch(url, { signal: AbortSignal.timeout(FETCH_TIMEOUT_MS) });
  if (!res.ok) {
    throw new Error(`HTTP ${res.status} ${res.statusText} for ${url}`);
  }
  return res.json();
}

async function geocodeZip(zip: string): Promise<WeatherLocation> {
  const data = (await fetchJson(`${ZIPPOPOTAM_URL}/${zip}`)) as {
    "post code"?: string;
    places?: Array<{
      "place name"?: string;
      latitude?: string;
      longitude?: string;
      "state abbreviation"?: string;
      state?: string;
    }>;
  };

  const place = data.places?.[0];
  const lat = place?.latitude ? Number.parseFloat(place.latitude) : NaN;
  const lon = place?.longitude ? Number.parseFloat(place.longitude) : NaN;
  if (!place || !Number.isFinite(lat) || !Number.isFinite(lon)) {
    throw new Error(`Could not resolve ZIP ${zip}`);
  }
  return {
    zip: data["post code"] ?? zip,
    place: place["place name"] ?? "",
    state: place["state abbreviation"] ?? place.state ?? "",
    lat,
    lon,
  };
}

// https://open-meteo.com/en/docs — WMO weather interpretation codes.
const WEATHER_CODE_LABELS: Record<number, string> = {
  0: "Clear sky",
  1: "Mainly clear",
  2: "Partly cloudy",
  3: "Overcast",
  45: "Fog",
  48: "Depositing rime fog",
  51: "Light drizzle",
  53: "Drizzle",
  55: "Heavy drizzle",
  56: "Light freezing drizzle",
  57: "Freezing drizzle",
  61: "Light rain",
  63: "Rain",
  65: "Heavy rain",
  66: "Light freezing rain",
  67: "Freezing rain",
  71: "Light snow",
  73: "Snow",
  75: "Heavy snow",
  77: "Snow grains",
  80: "Rain showers",
  81: "Heavy rain showers",
  82: "Violent rain showers",
  85: "Snow showers",
  86: "Heavy snow showers",
  95: "Thunderstorm",
  96: "Thunderstorm with hail",
  99: "Thunderstorm with heavy hail",
};

function describeWeatherCode(code: unknown): string {
  if (typeof code !== "number") return "Unknown";
  return WEATHER_CODE_LABELS[code] ?? `Weather code ${code}`;
}

interface OpenMeteoResponse {
  current?: {
    time?: string;
    temperature_2m?: number;
    apparent_temperature?: number;
    relative_humidity_2m?: number;
    wind_speed_10m?: number;
    wind_gusts_10m?: number;
    precipitation?: number;
    weather_code?: number;
    is_day?: number;
  };
  daily?: {
    time?: string[];
    weather_code?: number[];
    temperature_2m_max?: number[];
    temperature_2m_min?: number[];
    precipitation_probability_max?: number[];
    precipitation_sum?: number[];
    sunrise?: string[];
    sunset?: string[];
  };
}

async function fetchForecast(
  location: WeatherLocation,
  days: number
): Promise<{ current: WeatherCurrent; daily: WeatherDaily[] }> {
  const params = new URLSearchParams({
    latitude: String(location.lat),
    longitude: String(location.lon),
    temperature_unit: "fahrenheit",
    wind_speed_unit: "mph",
    precipitation_unit: "inch",
    timezone: "auto",
    current: [
      "temperature_2m",
      "apparent_temperature",
      "relative_humidity_2m",
      "wind_speed_10m",
      "wind_gusts_10m",
      "precipitation",
      "weather_code",
      "is_day",
    ].join(","),
    daily: [
      "weather_code",
      "temperature_2m_max",
      "temperature_2m_min",
      "precipitation_probability_max",
      "precipitation_sum",
      "sunrise",
      "sunset",
    ].join(","),
    forecast_days: String(Math.min(7, Math.max(1, days))),
  });

  const data = (await fetchJson(`${OPEN_METEO_FORECAST_URL}?${params.toString()}`)) as OpenMeteoResponse;

  const c = data.current ?? {};
  const current: WeatherCurrent = {
    time: c.time ?? "",
    temp_f: c.temperature_2m ?? Number.NaN,
    apparent_f: c.apparent_temperature ?? Number.NaN,
    humidity_pct: c.relative_humidity_2m ?? Number.NaN,
    wind_mph: c.wind_speed_10m ?? Number.NaN,
    wind_gusts_mph: c.wind_gusts_10m ?? Number.NaN,
    precipitation_in: c.precipitation ?? 0,
    weather: describeWeatherCode(c.weather_code),
    is_day: c.is_day === 1,
  };

  const d = data.daily ?? {};
  const times = d.time ?? [];
  const daily: WeatherDaily[] = times.map((date, i) => ({
    date,
    weather: describeWeatherCode(d.weather_code?.[i]),
    high_f: d.temperature_2m_max?.[i] ?? Number.NaN,
    low_f: d.temperature_2m_min?.[i] ?? Number.NaN,
    precip_chance_pct: d.precipitation_probability_max?.[i] ?? 0,
    precip_sum_in: d.precipitation_sum?.[i] ?? 0,
    sunrise: d.sunrise?.[i] ?? "",
    sunset: d.sunset?.[i] ?? "",
  }));

  return { current, daily };
}

export async function getWeatherForZip(
  zip: string,
  days: number
): Promise<WeatherResultPayload | WeatherErrorPayload> {
  const normalized = normalizeZip(zip);
  if (!normalized) {
    return { error: "Invalid ZIP code (expected 5 US digits)", zip };
  }
  try {
    const location = await geocodeZip(normalized);
    const { current, daily } = await fetchForecast(location, days);
    return { units: "imperial", location, current, daily };
  } catch (err) {
    return {
      error: err instanceof Error ? err.message : String(err),
      zip: normalized,
    };
  }
}

File: src-tauri/src/weather.rs
//! Ambient current weather for the desktop compose splash (US ZIP → Open-Meteo).

use std::time::Duration;

use reqwest::Client;
use serde::Serialize;
use serde_json::{json, Value};

use crate::settings;
use crate::storage::WriteChains;

const ZIPPOPOTAM_URL: &str = "https://api.zippopotam.us/us";
const OPEN_METEO_FORECAST_URL: &str = "https://api.open-meteo.com/v1/forecast";
const DEFAULT_ZIP: &str = "12528";

#[derive(Debug, Clone)]
struct WeatherLocation {
    zip: String,
    place: String,
    state: String,
    lat: f64,
    lon: f64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CurrentWeather {
    pub temp_f: f64,
    pub place: String,
    pub state: String,
    pub zip: String,
    pub weather: String,
    /// Quiet compose-corner label, e.g. `72° · Highland`.
    pub label: String,
}

pub fn normalize_zip(raw: &str) -> Option<String> {
    let digits: String = raw.chars().filter(|c| c.is_ascii_digit()).collect();
    if digits.len() < 5 {
        return None;
    }
    Some(digits.chars().take(5).collect())
}

fn describe_weather_code(code: i64) -> String {
    match code {
        0 => "Clear sky".into(),
        1 => "Mainly clear".into(),
        2 => "Partly cloudy".into(),
        3 => "Overcast".into(),
        45 | 48 => "Fog".into(),
        51 | 53 | 55 | 56 | 57 => "Drizzle".into(),
        61 | 63 | 65 | 66 | 67 | 80 | 81 | 82 => "Rain".into(),
        71 | 73 | 75 | 77 | 85 | 86 => "Snow".into(),
        95 | 96 | 99 => "Thunderstorm".into(),
        other => format!("Weather code {other}"),
    }
}

async fn fetch_json(client: &Client, url: &str) -> Result<Value, String> {
    let response = client
        .get(url)
        .timeout(Duration::from_secs(15))
        .send()
        .await
        .map_err(|e| e.to_string())?;
    if !response.status().is_success() {
        return Err(format!(
            "HTTP {} {}",
            response.status(),
            response.status().canonical_reason().unwrap_or("error")
        ));
    }
    response.json().await.map_err(|e| e.to_string())
}

async fn geocode_zip(client: &Client, zip: &str) -> Result<WeatherLocation, String> {
    let data = fetch_json(client, &format!("{ZIPPOPOTAM_URL}/{zip}")).await?;
    let place = data
        .get("places")
        .and_then(|v| v.as_array())
        .and_then(|arr| arr.first())
        .ok_or_else(|| format!("Could not resolve ZIP {zip}"))?;
    let lat = place
        .get("latitude")
        .and_then(|v| v.as_str())
        .and_then(|s| s.parse::<f64>().ok())
        .ok_or_else(|| format!("Could not resolve ZIP {zip}"))?;
    let lon = place
        .get("longitude")
        .and_then(|v| v.as_str())
        .and_then(|s| s.parse::<f64>().ok())
        .ok_or_else(|| format!("Could not resolve ZIP {zip}"))?;
    Ok(WeatherLocation {
        zip: data
            .get("post code")
            .and_then(|v| v.as_str())
            .unwrap_or(zip)
            .to_string(),
        place: place
            .get("place name")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string(),
        state: place
            .get("state abbreviation")
            .or_else(|| place.get("state"))
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string(),
        lat,
        lon,
    })
}

async fn fetch_current_temp(
    client: &Client,
    location: &WeatherLocation,
) -> Result<(f64, String), String> {
    let url = format!(
        "{OPEN_METEO_FORECAST_URL}?latitude={}&longitude={}&temperature_unit=fahrenheit&timezone=auto&current=temperature_2m,weather_code",
        location.lat, location.lon
    );
    let data = fetch_json(client, &url).await?;
    let current = data.get("current").cloned().unwrap_or_else(|| json!({}));
    let temp_f = current
        .get("temperature_2m")
        .and_then(|v| v.as_f64())
        .ok_or_else(|| "Missing temperature from forecast".to_string())?;
    let weather = describe_weather_code(
        current
            .get("weather_code")
            .and_then(|v| v.as_i64())
            .unwrap_or(-1),
    );
    Ok((temp_f, weather))
}

fn format_label(temp_f: f64, place: &str) -> String {
    let temp = format!("{}°", temp_f.round() as i64);
    let place = place.trim();
    if place.is_empty() {
        temp
    } else {
        format!("{temp} · {place}")
    }
}

pub async fn current_for_zip(zip: &str) -> Result<CurrentWeather, String> {
    let normalized =
        normalize_zip(zip).ok_or_else(|| "Invalid ZIP code (expected 5 US digits)".to_string())?;
    let client = Client::new();
    let location = geocode_zip(&client, &normalized).await?;
    let (temp_f, weather) = fetch_current_temp(&client, &location).await?;
    Ok(CurrentWeather {
        temp_f,
        place: location.place.clone(),
        state: location.state,
        zip: location.zip,
        weather,
        label: format_label(temp_f, &location.place),
    })
}

pub async fn current_from_settings(chains: &WriteChains) -> Result<CurrentWeather, String> {
    let settings = settings::get_settings(chains).await;
    let zip = settings
        .get("weather")
        .and_then(|v| v.get("defaultZip"))
        .and_then(|v| v.as_str())
        .unwrap_or(DEFAULT_ZIP)
        .trim();
    let zip = if zip.is_empty() { DEFAULT_ZIP } else { zip };
    current_for_zip(zip).await
}

#[tauri::command(rename_all = "camelCase")]
pub async fn weather_get_current(
    state: tauri::State<'_, crate::memory::AppState>,
) -> Result<CurrentWeather, String> {
    current_from_settings(&state.write_chains).await
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalize_zip_takes_first_five_digits() {
        assert_eq!(normalize_zip("12528").as_deref(), Some("12528"));
        assert_eq!(normalize_zip("12528-1234").as_deref(), Some("12528"));
        assert_eq!(normalize_zip("abc"), None);
        assert_eq!(normalize_zip("123"), None);
    }

    #[test]
    fn format_label_includes_place_when_present() {
        assert_eq!(format_label(72.4, "Highland"), "72° · Highland");
        assert_eq!(format_label(72.4, "  "), "72°");
    }
}
